### ISSUE TITLE
Docker Composer version 2.0 vs version 1.0

### ISSUE THREAD

**Issue by bergtwvd** [weaveworks/weave#2402]
Versions:
- Docker 1.11.2
- Docker Compose 1.7.1
- Weave 1.6.0

Situation:
- 2 Docker Hosts, Weave network between them
- DOCKER_HOST points to Weave Proxy
- Host 1: Compose file with container services defined, KV store running
- Host 2: Individual Docker container service on host 2
- All services use multicast

Test:
- when I use a version 1 format Compose file on Host 1 and start the individual container on host 2 they all see each other; there is an eth0 and ethwe interface in each container
- when I use a version 2 format Compose file (by just adding the version 2 line, and a services section at the top), Docker Compose creates a bridge that the containers on host 1 use to connect among each other; there is however no ethwe network interface in the started containers

- when I unset DOCKER_HOST on host 1, and add to the version 2 format Compose file a network as below, then a docker_gwbridge bridge network is created and a weave network is created. Each container has an ethwe0  and eth1 interface. However, the containers use the docker_gwbridge bridge network to (try to) connect to each other (anyway, they use an IP address from that network) and do not use the weave network. And the containers do not even see each other on docker_gwbridge network because of the option  "com.docker.network.bridge.enable_icc": "false".
```
networks:
 default:
  driver: weave
```

In conclusion, for multicast I can only use the version 1 Compose File and the Weave Proxy.
I cannot use the version 2 Compose file, and not use the version 2 Weave network driver.

**Comment by bboreham** [comment]
Do you have a KV store configured? (Doesn't sound like it, but the diagnosis would be different in that case)

**Comment by bboreham** [comment]
Oh, sorry, you said you had a KV store.

**Comment by bergtwvd** [comment]
I have the same problem when I run stand-alone on host 1 (with Weave and KV set up for host-1).
There appears to be a big diff between V1 and V2 compose file handling.

**Comment by bboreham** [comment]
OK, I tried to recreate your situation. I created a `docker-compose.yml` file as follows:

````
version: '2'

services:
  netcat:
    image: alpine
    command: nc -l -p 1234
networks:
  default:
    driver: weave
````

I launch Weave Net and start my service:

````
$ weave launch
$ docker-compose up -d
Creating network "bryan_default" with driver "weave"
Creating bryan_netcat_1
````

Then my netcat container has an `ethwe` interface:

````
$ docker exec bryan_netcat_1 ip addr show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
37: ethwe0@if38: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1410 qdisc noqueue state UP 
    link/ether 1a:23:bc:c6:3d:57 brd ff:ff:ff:ff:ff:ff
    inet 10.0.0.2/24 scope global ethwe0
       valid_lft forever preferred_lft forever
    inet6 fe80::1823:bcff:fec6:3d57/64 scope link 
       valid_lft forever preferred_lft forever
39: eth1@if40: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP 
    link/ether 02:42:ac:12:00:02 brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.2/16 scope global eth1
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe12:2/64 scope link 
       valid_lft forever preferred_lft forever
````

So, if you try the same steps, do you still not get an `ethwe` interface?

Please supply the output of `docker info`, `docker network ls` and `docker logs weaveplugin`.

**Comment by rade** [comment]
> I cannot use the version 2 Compose file

Yes, you can. See [the compose file in our guide](https://github.com/weaveworks/guides/blob/master/weave-and-docker-platform/app/docker-compose.yml). That is using the proxy, e.g. you can run it with

```
$ weave launch
$ eval $(weave env)
$ docker-compose up -d
```

> and not use the version 2 Weave network driver.

Yes, you can, but you must create custom networks, not just use the default 'weave' network. See [the compose file in our demo app](https://github.com/weaveworks/weaveDemo/blob/master/docker-compose.yml). You *must not* use the proxy in this mode (so *don't* run the 'eval' in the above).

Note that this will end up using docker's DNS, not weaveDNS.

Finally, you can use v2 syntax with the 'weave' network by making sure each service has a `dns` and `dns_search` stanza that reflects the output of `weave dns-args`.

This is all a bit crazy and should be much better explained. See weaveworks/guides#187.

**Comment by bergtwvd** [comment]
I have created a MC Test container, which is a simplified form for the apps that I have. You can find it at https://hub.docker.com/r/bergtwvd/po-mctest

Attached two docker compose test files:
- weavebridge.yml
- weaveoverlay.yml

**weavebridge** should be used in combination with the weave proxy.
**weaveoverlay** should be used in combination with the weave driver.

weavebridge works as expected.

For weaveoverlay the containers connect to the docker_gwbridge network and do not see each other at all. I would have expected they find eacht other on the weave network.

**Comment by rade** [comment]
> For weaveoverlay the containers connect to the docker_gwbridge network and do not see each other at all.

So name resolution inside the containers is returning an IP on the docker_gwbridge network? That would indicate a bug (or at least surprising behaviour) in docker's DNS resolver.

**Comment by bergtwvd** [comment]
Not sure about the "name resolution", since none of the containers is addressing any other one by name. The containers just discover each other over the network via the multi cast protocol.

**Comment by rade** [comment]
Multicast must be explicitly enabled for custom networks. See bottom of https://www.weave.works/docs/net/latest/plugin/plugin-how-it-works/ (and iirc there is a way to supply network options in compose)

**Comment by bergtwvd** [comment]
This appears hopeful.

How do I set this in the Docker Compose file.

Just did:
```
networks:
 training:
  driver: weave
  driver_opts:
   works.weave.multicast: "1"
```

Without effect though.

How do I set this option?

**Comment by rade** [comment]
> Without effect though.

How do you know? Did you check the route table in the containers?

> How do I set this option?

Your guess is as good as mine. What you wrote looks plausible.

**Comment by rade** [comment]
> What you wrote looks plausible.

Actually no, it needs to be a boolean.

**Comment by rade** [comment]
> Actually no, it needs to be a boolean.

Scratch that; we parse "1" as a boolean.

**Comment by bergtwvd** [comment]
On the blog page referred to there is also this note:

> Note: If you connect a container to multiple Weave networks, at most one of them can have the multicast route enabled. The weave network created when the plugin is first launched has the multicast option turned on, but for any networks you create it defaults to off.

Not sure how to interpret this. In view I am creating one network, should MC should be turned on already ...?

**Comment by rade** [comment]
> On the blog page referred to

What blog paged? Referred to from where?

>  In my view I am creating one network, should MC should be turned on already ...?

The behaviour changed in 1.6.0. I see we failed to point that out in the release notes.

**Comment by bergtwvd** [comment]
Check. A route has been added:

```
docker@docker-1:/c/Users/bergtwvd/docker/dockerfiles/build-po-mctest$ docker exec 1736deecaa8f route
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
default         172.17.0.1      0.0.0.0         UG    0      0        0 eth1
10.0.0.0        *               255.255.255.0   U     0      0        0 ethwe0
172.17.0.0      *               255.255.0.0     U     0      0        0 eth1
224.0.0.0       *               240.0.0.0       U     0      0        0 ethwe0
```

However, when I run weaveoverlay.yml the containers still pick 172.17.0.* addresses and do not see each other.

MC address 239.20.9.13 is used, which should route to ethwe0 ...

**Comment by rade** [comment]
> MC address 239.20.9.13 is used

Perhaps the *source* address is wrong?

**Comment by bergtwvd** [comment]
>    On the blog page referred to

>What blog paged? Referred to from where?

I was looking at page: https://www.weave.works/docs/net/latest/plugin/plugin-how-it-works, at the bottom. I suppose this is a logical restriction.

**Comment by bergtwvd** [comment]
> Perhaps the source address is wrong?

I am not a network specialist. In this case what is the source address used, and what should it be?

**Comment by rade** [comment]
You'd want the weave IP to be the source address. IIRC it's something that the app can do when creating the socket.

But first check that is actually the problem; run a packet capture on the ethwe0 interface of the multicast senders and check the source and destination addresses of the multicast packets.

**Comment by bergtwvd** [comment]
I force the application to use the ethwe0 interface to bind to, so the "source address" is now 10.0.0.0/24.
No different effect though.
What should I do to run the packet capture?

**Comment by rade** [comment]
> What should I do to run the packet capture?

Run tcpdump against the interface. I can't remember the necessary command line params offhand.

**Comment by bboreham** [comment]
I updated the 1.6.0 release notes to call attention to this change.

The compose file syntax to turn on the Weave Net multicast route is:

````
networks:
 training:
  driver: weave
  driver_opts:
      works.weave.multicast: "1"
````

> the containers still pick 172.17.0.* addresses

@bergtwvd what logic are the containers using to decide which interface to bind to?  Often you have to specify this explicitly to guarantee the desired result.

You can disable the docker-gwbridge interface by declaring the Weave network as "internal".  This cannot be done in a compose file, but you can pre-create the network:

    docker network create --internal --driver=weave --opt works.weave.multicast=true training

then refer to it thus:

````
networks:
 training:
  external: true
````

I did this with your `weaveoverlay.yml` example and saw in the logs "`physical address=10.0.0.4:39360`", which I think is what you were looking for.

**Comment by bergtwvd** [comment]
I repeated your suggestions, and now the applications see each other. You can recognise this in the output by the lines:

```
mctest3_1  | ERROR [main] portico.lrc.jgroups: FAILURE createFederation: already exists, name=TheWorld
mctest3_1  | Jun 27, 2016 10:06:47 PM mctest.Federate connectAndJoin
mctest3_1  | INFO: Didn't create federation, it already existed
mctest2_1  | ERROR [main] portico.lrc.jgroups: FAILURE createFederation: already exists, name=TheWorld
mctest2_1  | Jun 27, 2016 10:06:47 PM mctest.Federate connectAndJoin
mctest2_1  | INFO: Didn't create federation, it already existed
```
(hinting that the two later starting app join an existing federation, created by the first app)

And with the statements:
```
mctest3_1  | Jun 27, 2016 10:06:48 PM mctest.Federate connectAndJoin
mctest3_1  | INFO: Joined Federation as me3
mctest3_1  | Jun 27, 2016 10:06:48 PM mctest.Federate publishAndSubscribe
mctest3_1  | INFO: Published and Subscribed
mctest3_1  | Jun 27, 2016 10:06:48 PM mctest.FederateAmbassador discoverObjectInstance
**mctest3_1  | INFO: DISCOVERED: MOM.Federate(me3)**
mctest3_1  | Jun 27, 2016 10:06:48 PM mctest.FederateAmbassador discoverObjectInstance
**mctest3_1  | INFO: DISCOVERED: MOM.Federate(me1)**
mctest3_1  | Jun 27, 2016 10:06:48 PM mctest.FederateAmbassador discoverObjectInstance
**mctest3_1  | INFO: DISCOVERED: MOM.Federate(me2)**
```

I am still puzzled why this does not work in the regular way.

**Comment by bergtwvd** [comment]
> the containers still pick 172.17.0.* addresses

>>@bergtwvd what logic are the containers using to decide which interface to bind to? Often you have to specify this explicitly to guarantee the desired result.

This logic is implemented in JGroups (out of my control). I have some JGroups options I can set to control a few things, such as the JGROUPS_UDP_BINDADDRESS. I set this (optionally) to the IP address of ethwe0. I tried this earlier and I noticed 10.0.0.* addresses were used, but still the containers did not see each other.

**Comment by bergtwvd** [comment]
As a sanity check, I did another try with setting JGROUPS_UDP_BINDADDRESS to the IP of ethwe0..... and it works now :+1:   I must have overlooked something, or just did not have my glasses on !

Here the updated docker file:
[weaveoverlay.zip](https://github.com/weaveworks/weave/files/335954/weaveoverlay.zip)

I also updated and pushed the new image that supports ENV PORTICO_LRCADAPTER.

When I set the ENV PORTICO_LRCADAPTER=ethwe0 in the Docker Compose file then (as part of the application start) JGROUPS_UDP_BINDADDRESS is set to the IP address of ethwe0, and the 10.0.0.* addresses are used.

So, this works now.

Lacking network knowledge: if the ip route is added for multicast, why is this not sufficient?

**Comment by bboreham** [comment]
For sending multicast packets, we set a route so that multicast packets will go via the `weave` bridge and not take the default route via Docker's `gw_bridge` interface.

For listening, if the code specifically binds to a different interface, the route does not come into play.  Multicast packets arriving via the `weave` bridge will be delivered to the `ethwe0` interface and not to the interface the program is listening on. 

I don't know JGroups, but looking at https://docs.jboss.org/jbossas/docs/Server_Configuration_Guide/4/html/ch19s07s07.html it says "JGroups will bind to the machine's default interface", and explains how you can override it, which you've done.

Sadly, networking has lots of gotchas.  It sounds like you made some progress, which is great.
